### PR TITLE
Add test for 2.11 and 2.13 version of ExplicitResultType

### LINKED ISSUE #998: Make ExplicitResultTypes work with more Scala versions
Unlike other rules, `ExplicitResultTypes` only works with Scala 2.12.10 at the moment because the rule interfaces directly with the presentation compiler. It would be great if we could add support for Scala 2.13 and 2.11 but doing so would require rethinking how rules are compiled/published/launched.

* should Scalafix rules be cross-built against multiple Scala versions if they want to use the presentation compiler API? This would require changing sbt-scalafix to launch the correct Scalafix version (2.11/2.12/2.13) depending on the project being fixed.
* should Scalafix provide an Java-API on top of the presentation compiler allowing 2.12-compiled rules to work with 2.13/2.11? We use this approach in Metals (the server runs in 2.12, but completions work with 2.11/2.12/2.13).

I'm inclined to do the second approach since I like that Scalafix rules can work regardless of the Scala version being fixed. Regardless of the approach we use, I estimate it will require non-trivial work to implement.

### PR DESCRIPTION
**Pull request by mlachkar**
closes #998

### CODE REVIEW

File: scalafix-tests/output/src/main/scala-2.13/tests/ExplicitResultTypesBug.scala [hunk 1]
```diff
@@ -0,0 +1,56 @@
+package tests
```
Comment (bjaglin) [inline scalafix-tests/output/src/main/scala-2.13/tests/ExplicitResultTypesBug.scala:1]: so this one doesn't work on scala 2.11?

File: scalafix-tests/output/src/main/scala-2.11/test/explicitResultTypes/ExplicitResultTypesInfix.scala [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package test.explicitResultTypes
```
Comment (bjaglin) [inline scalafix-tests/output/src/main/scala-2.11/test/explicitResultTypes/ExplicitResultTypesInfix.scala:1]: so this one doesn't work yet on scala 2.13?

File: scalafix-tests/input/src/main/scala/test/explicitResultTypes/Rename.scala [hunk 3]
```diff
@@ -3,11 +3,11 @@ rules = ExplicitResultTypes
 */
 package test.explicitResultTypes
 
-import scala.collection.{Seq => SSeq}
+import scala.collection.immutable.{List => LList}
```
Comment (bjaglin) [inline scalafix-tests/input/src/main/scala/test/explicitResultTypes/Rename.scala:6]: why this change?

File: scalafix-tests/input/src/main/scala/tests/ExplicitResultTypesRefinement.scala [hunk 4]
```diff
@@ -27,11 +27,15 @@ object ExplicitResultTypesRefinement {
...
-  //  }
+  val access = new Serializable {
+    private val results: List[Int] = List.empty
+    protected val results2: List[Int] = List.empty
+  }
+  val product = new Product {
+    override def productArity: Int = ???
+    override def productElement(n: Int): Any = ???
+    override def canEqual(that: Any): Boolean = ???
+  }
```
Comment (bjaglin) [inline scalafix-tests/input/src/main/scala/tests/ExplicitResultTypesRefinement.scala:38]: maybe you can add a `new Serializable with Product` since the fix seems to be related to that?

File: scalafix-rules/src/main/scala/scalafix/internal/rule/CompilerTypePrinter.scala [hunk 5]
```diff
@@ -313,9 +313,19 @@ class CompilerTypePrinter(g: ScalafixGlobal, config: ExplicitResultTypesConfig)(
...
-          }
+          val productRootClass = definitions.ProductClass.seq
+            .toSet[Symbol] + definitions.ProductRootClass
+          val serializableClass = definitions.SerializableClass
+          val strippedParents =
+            if (productRootClass
+                .intersect(parents.map(_.typeSymbol).toSet)
+                .nonEmpty && parents
+                .map(_.typeSymbol)
+                .contains(serializableClass)) {
```
Comment (bjaglin) [inline scalafix-rules/src/main/scala/scalafix/internal/rule/CompilerTypePrinter.scala:324]: could you extract `val parentsClass = parents.map(_.typeSymbol).toSet` to make the condition more readable?

File: scalafix-tests/input/src/main/scala/tests/ExplicitResultTypesRefinement.scala (hunk 4 above)
Comment (mlachkar) [reply scalafix-tests/input/src/main/scala/tests/ExplicitResultTypesRefinement.scala:38]: good Idea.

File: scalafix-rules/src/main/scala/scalafix/internal/rule/CompilerTypePrinter.scala (hunk 5 above)
Comment (mlachkar) [reply scalafix-rules/src/main/scala/scalafix/internal/rule/CompilerTypePrinter.scala:324]: yes.

File: scalafix-tests/output/src/main/scala-2.11/test/explicitResultTypes/ExplicitResultTypesInfix.scala (hunk 2 above)
Comment (mlachkar) [reply scalafix-tests/output/src/main/scala-2.11/test/explicitResultTypes/ExplicitResultTypesInfix.scala:1]: So this one doesn't find the same types in 2.11 (compared to 2.12 and 2.13).
On 2.12 and 2.13, there is still an output test that looks like that
```
object ExplicitResultTypesInfix {
  val ab: Int \/ String = null.asInstanceOf[Int \/ String]
  def foo: Int \/ String = identity(ab)

  case class :+:[T1, T2]()
  def bar: Double :+: Long = :+:[Double, Long]()
}
```

File: scalafix-tests/output/src/main/scala-2.13/tests/ExplicitResultTypesBug.scala (hunk 1 above)
Comment (mlachkar) [reply scalafix-tests/output/src/main/scala-2.13/tests/ExplicitResultTypesBug.scala:1]: I forgot this. I will check

File: scalafix-tests/input/src/main/scala/test/explicitResultTypes/Rename.scala (hunk 3 above)
Comment (mlachkar) [reply scalafix-tests/input/src/main/scala/test/explicitResultTypes/Rename.scala:6]: scala.Seq is now an alias for scala.collection.immutable.Seq[A] (instead of scala.collection.Seq[A]) in 2.13.
So I would need different output tests. The goal is to test the Rename ... this is why I did this

File: scalafix-tests/output/src/main/scala-2.11/test/explicitResultTypes/ExplicitResultTypesInfix.scala (hunk 2 above)
Comment (bjaglin) [reply scalafix-tests/output/src/main/scala-2.11/test/explicitResultTypes/ExplicitResultTypesInfix.scala:1]: makes sense, sorry I ad missed the 3 different outputs!

File: scalafix-tests/output/src/main/scala/test/explicitResultTypes/Rename.scala [hunk 6]
```diff
@@ -1,11 +1,11 @@
 
 package test.explicitResultTypes
 
-import scala.collection.{Seq => SSeq}
+import scala.collection.immutable.{List => LList}
 import java.lang.{Boolean => JBoolean}
 
 object Rename {
-  type Seq = Int
-  def foo(a: Int*): SSeq[Int] = identity(a)
```
Comment (bjaglin) [inline scalafix-tests/output/src/main/scala/test/explicitResultTypes/Rename.scala:9]: I haven't looked at the git history, but maybe there was something with varargs/identity that this was specifically testing? In that case, maybe it's better to go for:
```
def foo(a: Int*): LList[Int] = identity(a.toList)
```

File: scalafix-tests/output/src/main/scala/test/explicitResultTypes/Rename.scala [hunk 7]
```diff
@@ -1,11 +1,11 @@
...
+  type List = Int
+  def foo(a: Int): LList[Int] = List(a)
```
Comment (bjaglin) [inline scalafix-tests/output/src/main/scala/test/explicitResultTypes/Rename.scala:9]: I haven't looked at the git history, but maybe there was something with varargs/identity that this was specifically testing? In that case, maybe it's better to go for:
```suggestion
  def foo(a: Int*): LList[Int] = identity(a.toList)
```

File: scalafix-rules/src/main/scala/scalafix/internal/rule/CompilerTypePrinter.scala [hunk 8]
```diff
@@ -222,6 +225,7 @@ class CompilerTypePrinter(g: ScalafixGlobal, config: ExplicitResultTypesConfig)(
 
   private def isPossibleSyntheticParent(tpe: Type): Boolean = {
     definitions.isPossibleSyntheticParent(tpe.typeSymbol) ||
+    CompilerCompat.serializableClass(g).toSet[Symbol](tpe.typeSymbol) ||
```
Comment (bjaglin) [inline scalafix-rules/src/main/scala/scalafix/internal/rule/CompilerTypePrinter.scala:228]: isn't that redundant with the check above?

File: scalafix-rules/src/main/scala/scalafix/internal/rule/CompilerTypePrinter.scala [hunk 9]
```diff
@@ -313,9 +317,19 @@ class CompilerTypePrinter(g: ScalafixGlobal, config: ExplicitResultTypesConfig)(
...
         case RefinedType(parents, _) =>
           // Remove redundant `Product with Serializable`, if possible.
-          val strippedParents = parents.filterNot { tpe =>
-            definitions.isPossibleSyntheticParent(tpe.typeSymbol)
-          }
+          val productRootClass = definitions.ProductClass.seq
+            .toSet[Symbol] + definitions.ProductRootClass
+          val serializableClass =
+            CompilerCompat.serializableClass(g).toSet[Symbol]
+          val parentSymbol = parents.map(_.typeSymbol).toSet
```
Comment (bjaglin) [inline scalafix-rules/src/main/scala/scalafix/internal/rule/CompilerTypePrinter.scala:324]:
```suggestion
          val parentSymbols = parents.map(_.typeSymbol).toSet
```

File: scalafix-rules/src/main/scala/scalafix/internal/rule/CompilerTypePrinter.scala (hunk 8 above)
Comment (mlachkar) [reply scalafix-rules/src/main/scala/scalafix/internal/rule/CompilerTypePrinter.scala:228]: not really... for scala 2.12 and 2.11, this method `definitions.isPossibleSyntheticParent` wasn't checking both Serializable traits.

Comment (bjaglin) [reply scalafix-rules/src/main/scala/scalafix/internal/rule/CompilerTypePrinter.scala:228]: got it, thanks!
